src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/FindAllReferences/RemoteFindAllReferencesService.cs

```diff
@@ -0,0 +1,116 @@
+        JsonSerializableDocumentId documentId,
+        Position position,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetImplementationAsync(context, position, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<RemoteResponse<SumType<VSInternalReferenceItem, LspLocation>[]?>> GetImplementationAsync(
+        RemoteDocumentContext context,
+        Position position,
+        CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        if (!codeDocument.Source.Text.TryGetAbsoluteIndex(position, out var hostDocumentIndex))
+        {
+            return NoFurtherHandling;
+        }
+
+        var positionInfo = GetPositionInfo(codeDocument, hostDocumentIndex, preferCSharpOverHtml: true);
+
+        if (positionInfo.LanguageKind is not RazorLanguageKind.CSharp)
+        {
+            return NoFurtherHandling;
+        }
+
+        // Finally, call into C#.
+        var generatedDocument = await context.Snapshot
+            .GetGeneratedDocumentAsync(cancellationToken)
+            .ConfigureAwait(false);
+
+        var results = await ExternalHandlers.FindAllReferences
+            .FindReferencesAsync(
+                RemoteWorkspaceAccessor.GetWorkspace(),
+                generatedDocument,
+                VsLspExtensions.ToLinePosition(positionInfo.Position),
+                _clientCapabilitiesService.ClientCapabilities.SupportsVisualStudioExtensions,
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        if (results is null and not [])
+        {
+            // C# didn't return anything, so we're done.
+            return NoFurtherHandling;
+        }
+
+        // Map the C# locations back to the Razor file.
+        foreach (var result in results)
+        {
+            var location = result.TryGetFirst(out var referenceItem)
+                ? referenceItem.Location
+                : result.Second;
+
+            if (location is null)
+            {
+                continue;
+            }
+
+            var (mappedUri, mappedRange) = await DocumentMappingService.MapToHostDocumentUriAndRangeAsync(context.Snapshot, location.Uri, location.Range.ToLinePositionSpan(), cancellationToken).ConfigureAwait(false);
+
+            if (referenceItem is not null)
+            {
+                // Indicates the reference item is directly available in the code
+                referenceItem.Origin = VSInternalItemOrigin.Exact;
+
+                // If we're going to change the Uri, then also override the file paths
+                if (mappedUri != location.Uri)
+                {
+                    referenceItem.DisplayPath = mappedUri.AbsolutePath;
+                    referenceItem.DocumentName = mappedUri.AbsolutePath;
```

Comment:
The docs here are not really clear, so I just looked at what was returned by Roslyn, which was full paths, and took that as gospel. I couldn't find any visual evidence in VS that the values are used. The docs say things like DisplayPath is for when "the actual path would be confusing", but whether that means we're doing the right thing because generated document paths are confusing, or the wrong thing because full paths are confusing, is entirely up to individual interpretation.